File: Day2/Mult.py
def products(nums):
    # Generate prefix products
    pre_prod = []#product before index
    for num in nums:
        if pre_prod: #check if pre_prod has elements
            pre_prod.append(pre_prod[-1] * num)#multiply last element of list
        else:
            pre_prod.append(num)

    # Generate suffix products
    after_prod = []
    for num in reversed(nums):
        if after_prod:
            after_prod.append(after_prod[-1] * num)
        else:
            after_prod.append(num)
    after_prod = list(reversed(after_prod))

    # Generate result
    result = []
    for i in range(len(nums)):
        if i == 0:
            result.append(after_prod[i + 1])
        elif i == len(nums) - 1:
            result.append(pre_prod[i - 1])
        else:
            result.append(pre_prod[i - 1] * after_prod[i + 1])
    return result

print(products(range(1,6)))

File: Day2/Mult.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class Mult{
    public static void main(String[] args){
        List<Integer> l=new ArrayList<>();
        for (int i = 1; i < 6; i++) {
            l.add(i);
        }
        l=products(l);
        System.out.println(l.toString());
        
    }
    public static List<Integer> products(List<Integer> l){
        int length=l.size();
        int[] preProd=new int[length];
        int[] afterProd=new int[length];
        preProd[0]=l.get(0);
        for (int i=1;i<length;i++) {
            preProd[i]=preProd[i-1]*l.get(i);
        }
        
        afterProd[length-1]=l.get(length-1);
        for(int i=length-2;i>=0;i--){
            afterProd[i]=afterProd[i+1]*l.get(i);
        }
        List<Integer> result=new ArrayList<>();
        for (int i = 0; i < length; i++) {
            if (i == 0){
                result.add(afterProd[i + 1]);
            }
            else if( i == length - 1){
                result.add(preProd[i - 1]);
            }
            else{
                result.add(preProd[i - 1] * afterProd[i + 1]);
            }
        }     
        return result;
    }
}